# **20BCE1205**

# **Shubham Ojha**

Q- Download a dataset, implement SVM and Logistic Regression classification algorithms. Compare the results

Note: Library functions can be used

In [ ]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
sns.set(color_codes=True)
%matplotlib inline

In [ ]:
data_frame = pd.read_csv("/content/lab7.csv")
data_frame.shape

(768, 9)

In [ ]:
data_frame.head(5)

num_preg  glucose_conc  diastolic_bp  skin_thickness  insulin   bmi  \
0         6           148            72              35        0  33.6   
1         1            85            66              29        0  26.6   
2         8           183            64               0        0  23.3   
3         1            89            66              23       94  28.1   
4         0           137            40              35      168  43.1   

   diab_pred  age  diabetes  
0      0.627   50      True  
1      0.351   31     False  
2      0.672   32      True  
3      0.167   21     False  
4      2.288   33      True

In [ ]:
data_frame.corr()

num_preg  glucose_conc  diastolic_bp  skin_thickness  \
num_preg        1.000000      0.129459      0.141282       -0.081672   
glucose_conc    0.129459      1.000000      0.152590        0.057328   
diastolic_bp    0.141282      0.152590      1.000000        0.207371   
skin_thickness -0.081672      0.057328      0.207371        1.000000   
insulin        -0.073535      0.331357      0.088933        0.436783   
bmi             0.017683      0.221071      0.281805        0.392573   
diab_pred      -0.033523      0.137337      0.041265        0.183928   
age             0.544341      0.263514      0.239528       -0.113970   
diabetes        0.221898      0.466581      0.065068        0.074752   

                 insulin       bmi  diab_pred       age  diabetes  
num_preg       -0.073535  0.017683  -0.033523  0.544341  0.221898  
glucose_conc    0.331357  0.221071   0.137337  0.263514  0.466581  
diastolic_bp    0.088933  0.281805   0.041265  0.239528  0.065068  
skin_thickness  0.436783  0.392573   0.183928 -0.113970  0.074752  
insulin         1.000000  0.197859   0.185071 -0.042163  0.130548  
bmi             0.197859  1.000000   0.140647  0.036242  0.292695  
diab_pred       0.185071  0.140647   1.000000  0.033561  0.173844  
age            -0.042163  0.036242   0.033561  1.000000  0.238356  
diabetes        0.130548  0.292695   0.173844  0.238356  1.000000

In [ ]:
data_frame.head(5)

num_preg  glucose_conc  diastolic_bp  skin_thickness  insulin   bmi  \
0         6           148            72              35        0  33.6   
1         1            85            66              29        0  26.6   
2         8           183            64               0        0  23.3   
3         1            89            66              23       94  28.1   
4         0           137            40              35      168  43.1   

   diab_pred  age  diabetes  
0      0.627   50      True  
1      0.351   31     False  
2      0.672   32      True  
3      0.167   21     False  
4      2.288   33      True

In [ ]:
diabetes_map = {True : 1, False : 0}
data_frame['diabetes'] = data_frame['diabetes'].map(diabetes_map)

In [ ]:
from sklearn.model_selection import train_test_split

feature_col_names = ['num_preg', 'glucose_conc', 'diastolic_bp', 'skin_thickness', 'insulin', 'bmi', 'diab_pred', 'age']
predicted_class_names = ['diabetes']

X = data_frame[feature_col_names].values
y = data_frame[predicted_class_names].values
split_test_size = 0.30

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=split_test_size, random_state=42) 

In [ ]:
trainval = (1.0 * len(X_train)) / (1.0 * len(data_frame.index))
testval = (1.0 * len(X_test)) / (1.0 * len(data_frame.index))

In [ ]:
print("# rows in dataframe {0}".format(len(data_frame)))
print("# rows missing glucose_conc: {0}".format(len(data_frame.loc[data_frame['glucose_conc'] == 0])))
print("# rows missing diastolic_bp: {0}".format(len(data_frame.loc[data_frame['diastolic_bp'] == 0])))
print("# rows missing thickness: {0}".format(len(data_frame.loc[data_frame['skin_thickness'] == 0])))
print("# rows missing insulin: {0}".format(len(data_frame.loc[data_frame['insulin'] == 0])))
print("# rows missing bmi: {0}".format(len(data_frame.loc[data_frame['bmi'] == 0])))
print("# rows missing diab_pred: {0}".format(len(data_frame.loc[data_frame['diab_pred'] == 0])))
print("# rows missing age: {0}".format(len(data_frame.loc[data_frame['age'] == 0])))

# rows in dataframe 768
# rows missing glucose_conc: 5
# rows missing diastolic_bp: 35
# rows missing thickness: 227
# rows missing insulin: 374
# rows missing bmi: 11
# rows missing diab_pred: 0
# rows missing age: 0


# **SVM**

In [ ]:
from sklearn.svm import SVC
svm_model = SVC(kernel='linear', C=1, random_state=42)

svm_model.fit(X_train, y_train.ravel())

SVC(C=1, kernel='linear', random_state=42)

In [ ]:
prediction_from_trained_data = svm_model.predict(X_train)

In [ ]:
from sklearn import metrics

accuracy = metrics.accuracy_score(y_train, prediction_from_trained_data)

print ("Accuracy of our SVM model is : {0:.4f}".format(accuracy))

Accuracy of our SVM model is : 0.7803


In [ ]:
svm_predict_test = svm_model.predict(X_test)

#get accuracy
svm_accuracy_testdata = metrics.accuracy_score(y_test, svm_predict_test)

#print accuracy
print ("Accuracy: {0:.4f}".format(svm_accuracy_testdata))

Accuracy: 0.7446


In [ ]:
print ("Confusion Matrix for Support Vector Amchine")

# labels for set 1=True to upper left and 0 = False to lower right
print ("{0}".format(metrics.confusion_matrix(y_test, svm_predict_test, labels=[1, 0])))

print ("")

print ("Classification Report\n")

# labels for set 1=True to upper left and 0 = False to lower right
print ("{0}".format(metrics.classification_report(y_test, svm_predict_test, labels=[1, 0])))

Confusion Matrix for Support Vector Amchine
[[ 50  30]
 [ 29 122]]

Classification Report

              precision    recall  f1-score   support

           1       0.63      0.62      0.63        80
           0       0.80      0.81      0.81       151

    accuracy                           0.74       231
   macro avg       0.72      0.72      0.72       231
weighted avg       0.74      0.74      0.74       231



# **Logistic** **Regression**

In [ ]:
from sklearn.linear_model import LogisticRegression

lr_model = LogisticRegression(C=0.7, random_state=42)
lr_model.fit(X_train, y_train.ravel())
lr_predict_test = lr_model.predict(X_test)

# training metrics
print ("Accuracy : {0:.4f}".format(metrics.accuracy_score(y_test, lr_predict_test)))

print ("Confusion Matrix")

print (metrics.confusion_matrix(y_test, lr_predict_test, labels=[1, 0]))

print ("")

print ("Classification Report")

print (metrics.classification_report(y_test, lr_predict_test, labels=[1, 0]))

Accuracy : 0.7403
Confusion Matrix
[[ 50  30]
 [ 30 121]]

Classification Report
              precision    recall  f1-score   support

           1       0.62      0.62      0.62        80
           0       0.80      0.80      0.80       151

    accuracy                           0.74       231
   macro avg       0.71      0.71      0.71       231
weighted avg       0.74      0.74      0.74       231



/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG,


In [ ]:
from sklearn.linear_model import LogisticRegressionCV

lr_cv_model = LogisticRegressionCV(n_jobs=-1, random_state=42, Cs=3, cv=10, refit=False, class_weight="balanced")  

# set number of jobs to -1 which uses all cores to parallelize
lr_cv_model.fit(X_train, y_train.ravel())

lr_cv_predict_test = lr_cv_model.predict(X_test)

# training metrics
print( "Accuracy: {0:.4f}".format(metrics.accuracy_score(y_test, lr_cv_predict_test)))
print (metrics.confusion_matrix(y_test, lr_cv_predict_test, labels=[1, 0])) 
print ("")
print ("Classification Report")
print (metrics.classification_report(y_test, lr_cv_predict_test, labels=[1,0]))

Accuracy: 0.7013
[[ 54  26]
 [ 43 108]]

Classification Report
              precision    recall  f1-score   support

           1       0.56      0.68      0.61        80
           0       0.81      0.72      0.76       151

    accuracy                           0.70       231
   macro avg       0.68      0.70      0.68       231
weighted avg       0.72      0.70      0.71       231



**Accuracy of SVM for test dataset is 0.7446 while for Logistic Regression is 0.7403. By using Logistic Regression cross-validation accuracy is 0.7013.
Therefore, SVM is better for our test dataset.**